### PR TITLE
Improvements to Basic_viewer

### PR DESCRIPTION
**Pull request by lrineau**
## Summary of Changes

This is a work-in-progress. I plan to submit improvement to the `Basic_viewer_qt`.

What I plan to do (with the help of @maxGimeno):
- [x] Move the definitions of shortcuts to the constructor of `Basic_viewer_qt` so that is can easily be modified in derived class.
- [x] Use a proper normal property map for `Surface_mesh` and `Polyhedron`, to avoid to recompute normals on the fly per vertex/face (that seems to be uneffective) 
- [x] Factorize the viewer for `Polyhedron` and `Surface_mesh`.
- [x] Turn viewer template classes into non-template classes.

## Release Management

* Affected package(s): BGL, GraphicsView, Polyhedron, Surface_mesh
* License and copyright ownership: unchanged, mainteance  by GeometryFactory

### CODE REVIEW

**Comment by gdamiand** [comment]
@lrineau How it is possible to remove template parameter for viewer classes?

**Comment by lrineau** [comment]
> @lrineau How it is possible to remove template parameter for viewer classes?

The idea is "type erasure". Here the type erasure is easy, because only `compute_elements()` depends on the actual type of mesh. One can type-erase that function using `std::function<void()>`. I am about to push a commit showing the solution.

**Comment by lrineau** [comment]
See https://github.com/CGAL/cgal/pull/4876/commits/b5dff8d3903d7fd6f6d6f9ae34c0c92966846c5c for the type erasure.

**Comment by lrineau** [comment]
@MaelRL @sloriot @afabri In this PR, I have improve `<CGAL/draw_surface_mesh.h>` and `<CGAL/draw_polyhedron.h>`:
  - factorized into a `<CGAL/draw_face_graph.h>`, using the `FaceGraph` concept,
  - I use dynamic property maps, and `<CGAL/Polygon_mesh_processing/compute_normal.h>` to speed-up the drawing.

Problem: now that creates a dependency to PMP.

I thought it was planned to move part of PMP to a LGPL part (maybe in `BGL/include`), to avoid such issues. Do we still plan that?

File: Surface_mesh/include/CGAL/draw_surface_mesh.h [hunk 1]
```diff
@@ -12,6 +12,9 @@
 #ifndef CGAL_DRAW_SURFACE_MESH_H
 #define CGAL_DRAW_SURFACE_MESH_H
 
+#include <CGAL/Polygon_mesh_processing/compute_normal.h>
+#include <CGAL/Dynamic_property_map.h>
```
Comment (sloriot) [inline Surface_mesh/include/CGAL/draw_surface_mesh.h:16]: Both include are useless AFAIS

**Comment by sloriot** [comment]
This is still planned. I don't think it is an issue. QGLViewer is GPL anyway.

File: BGL/include/CGAL/draw_face_graph.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    using Kernel = typename CGAL::Kernel_traits<Point>::Kernel;
+    using Vector = typename Kernel::Vector_3;
+
+    auto vnormals = get(CGAL::dynamic_vertex_property_t<Vector>(), sm);
+    {
+      // temporary face property map needed by `compute_normals`
+      auto fpm = get(CGAL::dynamic_face_property_t<Vector>(), sm);
+
+      CGAL::Polygon_mesh_processing::compute_normals(sm, vnormals, fpm);
+    }
```
Comment (lrineau) [inline BGL/include/CGAL/draw_face_graph.h:115]: As a reply to https://github.com/CGAL/cgal/pull/4876#discussion_r461545678:
@sloriot This is where I use `<CGAL/Dynamic_property_map.h>` and `<CGAL/Polygon_mesh_processing/compute_normal.h>`.

File: Surface_mesh/include/CGAL/draw_surface_mesh.h (hunk 1 above)
Comment (lrineau) [reply Surface_mesh/include/CGAL/draw_surface_mesh.h:16]: See https://github.com/CGAL/cgal/pull/4876#discussion_r461551666.

**Comment by lrineau** [comment]
> This is still planned. I don't think it is an issue. QGLViewer is GPL anyway.

The issue is that, in this PR, `Surface_mesh` now depends on PMP, because `Surface_mesh/include/CGAL/draw_surface_mesh.h` depends on PMP.

And BGL also depends on PMP because of `BGL/include/CGAL/draw_face_graph.h`.

See https://travis-ci.com/github/CGAL/cgal/jobs/365357831#L8200 (I do not understand why Travis does not see the dependency of BGL to PMP).

File: BGL/include/CGAL/draw_face_graph.h (hunk 2 above)
Comment (sloriot) [reply BGL/include/CGAL/draw_face_graph.h:115]: Which are already included in this header...

File: Surface_mesh/include/CGAL/draw_surface_mesh.h (hunk 1 above)
Comment (lrineau) [reply Surface_mesh/include/CGAL/draw_surface_mesh.h:16]: Oh, right. They are not needed in `CGAL/draw_surface_mesh.h`. My fault. That was an intermediate step of the refactoring.

**Comment by lrineau** [comment]
> This is still planned. I don't think it is an issue. QGLViewer is GPL anyway.

The issue is that is creates a new dependency from BGL to PMP.

**Comment by sloriot** [comment]
I replaced the call by a naive implementation that should be sufficient for display

**Comment by sloriot** [comment]
I don't think this is an issue as the code is only used for display. If you want to store normal per face, you'll then have to make a special case for non-triangular faces.

**Comment by maxGimeno** [comment]
Conflicts

**Comment by maxGimeno** [comment]
Trailing whitespaces : https://travis-ci.com/github/CGAL/cgal/jobs/395090185#L3175

**Comment by lrineau** [comment]
In a private conversion, @gdamiand approuved the modifications.

**Comment by lrineau** [comment]
> @MaelRL @sloriot @afabri In this PR, I have improve `<CGAL/draw_surface_mesh.h>` and `<CGAL/draw_polyhedron.h>`:
> 
>     * factorized into a `<CGAL/draw_face_graph.h>`, using the `FaceGraph` concept,
> 
>     * I use dynamic property maps, and `<CGAL/Polygon_mesh_processing/compute_normal.h>` to speed-up the drawing.
> 
> 
> Problem: now that creates a dependency to PMP.
> 
> I thought it was planned to move part of PMP to a LGPL part (maybe in `BGL/include`), to avoid such issues. Do we still plan that?

Now we have that license issue. Can we create a sub-part of PMP that would be LGPL?

**Comment by sloriot** [comment]
Why are you ignoring my comment?

**Comment by lrineau** [comment]
> Why are you ignoring my comment?

It was not ignored. See https://github.com/CGAL/cgal/pull/4876#issuecomment-678637035. Then [you said](https://github.com/CGAL/cgal/pull/4876#issuecomment-678951299): 
> I don't think this is an issue as the code is only used for display. If you want to store normal per face, you'll then have to make a special case for non-triangular faces.

I will bench the difference.

**Comment by lrineau** [comment]
> > Why are you ignoring my comment?
> 
> It was not ignored. See [#4876 (comment)](https://github.com/CGAL/cgal/pull/4876#issuecomment-678637035). Then [you said](https://github.com/CGAL/cgal/pull/4876#issuecomment-678951299):
> 
> > I don't think this is an issue as the code is only used for display. If you want to store normal per face, you'll then have to make a special case for non-triangular faces.
> 
> I will bench the difference.

You were right, and even on a big mesh with 1.5M vertices, I have not seen the difference in time.

**Comment by lrineau** [comment]
@maxGimeno This PR is now fine, and ready to be tested, in the next batch of PRs.

**Comment by maxGimeno** [comment]
Errors : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-26/Polyhedron_Examples/TestReport_lrineau_Fedora-Release.gz

**Comment by maxGimeno** [comment]
Another warning: https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.2-Ic-26/Surface_mesh_Examples/TestReport_lrineau_ArchLinux-clang-Release.gz

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.2-Ic-27.shtml

**Pull request merged**